report/renderer/dom.js

```diff
@@ -102,36 +105,23 @@ export class DOM {
   }
 
   /**
-   * @param {string} selector
-   * @param {ParentNode} context
-   * @return {!DocumentFragment} A clone of the template content.
-   * @throws {Error}
+   * @param {Parameters<typeof createComponent>[1]} componentName
+   * @return {!DocumentFragment} A clone of the cached component.
    */
-  cloneTemplate(selector, context) {
-    const template = /** @type {?HTMLTemplateElement} */ (context.querySelector(selector));
-    if (!template) {
-      throw new Error(`Template not found: template${selector}`);
-    }
-
-    const clone = this._document.importNode(template.content, true);
-
-    // Prevent duplicate styles in the DOM. After a template has been stamped
-    // for the first time, remove the clone's styles so they're not re-added.
-    if (template.hasAttribute('data-stamped')) {
-      this.findAll('style', clone).forEach(style => style.remove());
+  createComponent(componentName) {
+    let component = this._componentCache.get(componentName);
```

Comment:
does cache and clone (vs just running the dom creation code again every time) significantly impact things?

Reply:
> does cache and clone (vs just running the dom creation code again every time) significantly impact things?

bump

Reply:
bump :)

Reply:
![image](https://user-images.githubusercontent.com/4071474/127702265-a999cbad-17e3-41a6-9d8c-5b18c6f769fd.png)

It saves about 350-550ms on my machine.

Interestingly caching `audit` is sometimes slower, but not often, and overall caching everything is faster.

As this is ~500ms of time before the first render, this is significant.


```
diff --git a/report/assets/standalone-template.html b/report/assets/standalone-template.html
index cca9e9913..27d52b9e9 100644
--- a/report/assets/standalone-template.html
+++ b/report/assets/standalone-template.html
@@ -34,6 +34,14 @@ limitations under the License.
   <script>window.__LIGHTHOUSE_JSON__ = %%LIGHTHOUSE_JSON%%;</script>
   <script>%%LIGHTHOUSE_JAVASCRIPT%%
   __initLighthouseReport__();
+  let total = 0;
+  for (const [componentName, {init, cached}] of Object.entries(window.timings)) {
+    let t = 0;
+    for (const c of cached) t += (init - c);
+    total += t;
+    console.log({componentName, init}, 'num cached:', cached.length, 'saved time:', t);
+  }
+  console.log(total);
   //# sourceURL=compiled-reportrenderer.js
   </script>
   <script>console.log('window.__LIGHTHOUSE_JSON__', __LIGHTHOUSE_JSON__);</script>
diff --git a/report/renderer/dom.js b/report/renderer/dom.js
index 94b5dd0fe..1438513c5 100644
--- a/report/renderer/dom.js
+++ b/report/renderer/dom.js
@@ -93,17 +93,23 @@ export class DOM {
    */
   createComponent(componentName) {
     let component = this._componentCache.get(componentName);
+
+    const now = performance.now();
+    window.timings = window.timings || {};
+
     if (component) {
       const cloned = /** @type {DocumentFragment} */ (component.cloneNode(true));
       // Prevent duplicate styles in the DOM. After a template has been stamped
       // for the first time, remove the clone's styles so they're not re-added.
       this.findAll('style', cloned).forEach(style => style.remove());
+      window.timings[componentName].cached.push(performance.now() - now);
       return cloned;
     }
 
     component = createComponent(this, componentName);
     this._componentCache.set(componentName, component);
     const cloned = /** @type {DocumentFragment} */ (component.cloneNode(true));
+    window.timings[componentName] = {init: performance.now() - now, cached: []};
     return cloned;
   }
 
```

Reply:
Something is a bit wonky with this version of a comparison. Running with and without a cache renders pretty indistinguishably on my machine (always ~250ms for `__initLighthouseReport__`). Not very elegant version:

```diff
diff --git a/report/renderer/dom.js b/report/renderer/dom.js
index 94b5dd0fe..a0e9e0f1d 100644
--- a/report/renderer/dom.js
+++ b/report/renderer/dom.js
@@ -31,8 +31,8 @@ export class DOM {
     this._document = document;
     /** @type {string} */
     this._lighthouseChannel = 'unknown';
-    /** @type {Map<string, DocumentFragment>} */
-    this._componentCache = new Map();
+    /** @type {Map<string, boolean>} */
+    this._componentUsed = new Map();
   }
 
   /**
@@ -92,19 +92,17 @@ export class DOM {
    * @return {!DocumentFragment} A clone of the cached component.
    */
   createComponent(componentName) {
-    let component = this._componentCache.get(componentName);
-    if (component) {
-      const cloned = /** @type {DocumentFragment} */ (component.cloneNode(true));
+    const component = createComponent(this, componentName);
+    const isUsed = this._componentUsed.get(componentName);
+    if (isUsed) {
       // Prevent duplicate styles in the DOM. After a template has been stamped
       // for the first time, remove the clone's styles so they're not re-added.
-      this.findAll('style', cloned).forEach(style => style.remove());
-      return cloned;
+      this.findAll('style', component).forEach(style => style.remove());
+    } else {
+      this._componentUsed.set(componentName, true);
     }
 
-    component = createComponent(this, componentName);
-    this._componentCache.set(componentName, component);
-    const cloned = /** @type {DocumentFragment} */ (component.cloneNode(true));
-    return cloned;
+    return component;
   }
 
   /**
```

Maybe because your comparison is missing out on repeated component building, so nothing gets amortized? You wouldn't get a ton of JIT benefits because it's a lot of calls to DOM bindings, but there's a lot of type feedback that could maybe help the 472 elementScreenshot calls vs that initial `init` call.

If it's a wash performance wise, then whatever you think is best SG.

Reply:
I measured the time for `__initLighthouseReport__` just now and the timings I saw suggest a ~100-200ms savings, although with the variance and limited samples it's hard to be sure. I'd rather keep it to be safe, it can't hurt performance and it's only a few lines in this function.